src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -41,62 +46,128 @@ def from_slice(cls, slice: Slice[Motor], ramp_up_time: float) -> _Trajectory:
         """
         slice_duration = error_if_none(slice.duration, "Slice must have a duration")
 
-        # Check if any gaps other than initial gap.
-        if any(slice.gap[1:]):
-            raise RuntimeError(
-                f"Cannot parse trajectory with gaps. Slice has gaps: {slice.gap}"
-            )
-
         scan_size = len(slice)
+        gaps: list[int] = np.where(slice.gap)[0].tolist()
+        gaps.append(len(slice))
         motors = slice.axes()
 
         positions: dict[Motor, npt.NDArray[np.float64]] = {}
         velocities: dict[Motor, npt.NDArray[np.float64]] = {}
 
         # Initialise arrays
-        positions = {motor: np.empty(2 * scan_size + 1, float) for motor in motors}
-        velocities = {motor: np.empty(2 * scan_size + 1, float) for motor in motors}
-        durations: npt.NDArray[np.float64] = np.empty(2 * scan_size + 1, float)
-        user_programs: npt.NDArray[np.int32] = np.ones(2 * scan_size + 1, float)
-        user_programs[-1] = 8
+        trajectory_size = ((2 * scan_size) + 1) + ((len(gaps) - 2) * 4)
+        positions = {motor: np.empty(trajectory_size, float) for motor in motors}
+        velocities = {motor: np.empty(trajectory_size, float) for motor in motors}
+        durations: npt.NDArray[np.float64] = np.empty(trajectory_size, float)
+        user_programs: npt.NDArray[np.int32] = np.ones(trajectory_size, float)
 
         # Ramp up time for start of collection window
         durations[0] = int(ramp_up_time / TICK_S)
-        # Half the time per point
-        durations[1:] = np.repeat(slice_duration / (2 * TICK_S), 2)
 
-        # Fill profile assuming no gaps
-        # Excluding starting points, we begin at our next frame
         half_durations = slice_duration / 2
+
+        # Precompute gaps
+        gap_points = {}
+        for gap in gaps[1:-1]:
+            # Get entry velocities, exit velocities, and distances across gap
+            start_velocities, end_velocities, distances = _get_start_and_end_velocities(
+                motors, slice, half_durations, gap
+            )
+
+            # Get velocity and time profiles across gap
+            time_arrays, velocity_arrays = _get_velocity_profile(
+                motors, motor_info, start_velocities, end_velocities, distances
+            )
+
+            start_positions = {motor: slice.upper[motor][gap - 1] for motor in motors}
```

Comment:
You could do this inside `_calculate_profile_from_velocities`. It isn't used anywhere else.